pkg/controllers/provisioning/scheduling/node.go

```diff
@@ -0,0 +1,173 @@
+/*
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package scheduling
+
+import (
+	"errors"
+
+	v1 "k8s.io/api/core/v1"
+
+	"github.com/aws/karpenter/pkg/apis/provisioning/v1alpha5"
+	"github.com/aws/karpenter/pkg/cloudprovider"
+	"github.com/aws/karpenter/pkg/utils/resources"
+)
+
+// Node is a set of constraints, compatible pods, and possible instance types that could fulfill these constraints. This
+// will be turned into one or more actual node instances within the cluster after bin packing.
+type Node struct {
+	Constraints         *v1alpha5.Constraints
+	InstanceTypeOptions []cloudprovider.InstanceType
+	Pods                []*v1.Pod
+}
+
+func NewNode(constraints *v1alpha5.Constraints, instanceTypes []cloudprovider.InstanceType, pods ...*v1.Pod) *Node {
+	n := &Node{Constraints: constraints.DeepCopy()}
+
+	provIts := constraints.Requirements.InstanceTypes()
+	for _, it := range instanceTypes {
+		// provisioner must list the instance type by name (defaults to all instance types)
+		if provIts.Has(it.Name()) {
+			included := false
+			// and the instance type must have some valid offering combination per the provisioner constraints
+			for _, off := range it.Offerings() {
+				if constraints.Requirements.Zones().Has(off.Zone) && constraints.Requirements.CapacityTypes().Has(off.CapacityType) {
+					included = true
+					break
+				}
+			}
+			if included {
+				n.InstanceTypeOptions = append(n.InstanceTypeOptions, it)
+			}
+		}
+	}
+	// technically we might create some invalid construct here if the pod can't be supported by any instance type
+	// but this is ok as we just won't have any valid instance types, this pod won't be schedulable and nothing else
+	// will be compatible with this node
+	for _, p := range pods {
+		n.Add(p)
+	}
+	return n
+}
+func (n Node) Compatible(pod *v1.Pod) error {
+	podRequirements := v1alpha5.NewPodRequirements(pod)
+	if err := n.Constraints.Requirements.Compatible(podRequirements); err != nil {
+		return err
+	}
+
+	// Ensure that at least one instance type of the instance types that we are already narrowed down to based on the
+	// existing pods can support the pod
+	for _, it := range n.InstanceTypeOptions {
+		if n.isPodCompatible(pod, it) {
+			return nil
+		}
+	}
+
+	return errors.New("no matching instance type found")
```

Comment:
The errors package functionality is now replaced by fmt.Errorf and %w: https://github.com/pkg/errors/issues/202
```suggestion
	return errors.New("no matching instance type found")
```

Reply:
Also -- would be great to wire through a reason, since this may be a bit hard to reason about for users.

Reply:
I think errors.New() is still preferred if you don't have a need for the Sprintf machinery and %w, I've got no other error to pass up. e.g. from stdlib 

```
package io
var EOF = errors.New("EOF")
```

This error is never presented to users, at least not currently.  Compatible could just be a boolean, but I made it return an error to be similar to other Compatible methods.